File: src/game/systems/trigger.lua
local Aspect = require 'ecs.aspect'
local System = require 'ecs.system'
local Player = require 'game.components.player'
local Trigger = require 'game.components.trigger'
local collision = require 'game.utils.collision'

local aspect = Aspect.new({Trigger})
local TriggerSystem = System:new('trigger', aspect)

function TriggerSystem:update(dt)
  for _, entity in pairs(self.entities) do
    local trigger = entity:as(Trigger)

    if (trigger.activated and not trigger.executed) then
      trigger.action()
      trigger.executed = true
    end
  end
end

function TriggerSystem:beginContact(a, b, contact)
  a = collision.entity(a)
  b = collision.entity(b)

  if a:has(Trigger) and b:has(Player) then
    a:as(Trigger).activated = true
  elseif b:has(Trigger) and a:has(Player) then
    b:as(Trigger).activated = true
  end
end

return TriggerSystem

File: packages/client/src/game/systems/Trigger.ts
import { System } from '../../ecs/System'
import { SystemFlag } from '../flags'
import { Aspect } from '../../ecs/Aspect'
import { check, hasComponent } from '../utils/collision'
import { Trigger } from '../components/Trigger'
import { Player } from '../components/Player'

export class TriggerSystem extends System {
  static _id = 'Trigger'
  _id = TriggerSystem._id

  static _flag = SystemFlag.Trigger
  _flag = TriggerSystem._flag

  static _aspect = new Aspect([Trigger])
  _aspect = TriggerSystem._aspect

  update = (dt: number) => {
    this.entities.forEach(entity => {
      const trigger = entity.as(Trigger)

      if (trigger && trigger.activated && !trigger.executed) {
        trigger.action()
        trigger.executed = true
      }
    })
  }

  beginContact = (a: Fixture, b: Fixture, contact: Contact) => {
    const result = check(a, b, [hasComponent(Trigger), hasComponent(Player)])

    if (!result) {
      return
    }

    const trigger = result[0].as(Trigger)

    if (trigger) {
      trigger.activated = true
    }
  }
}
